tutorials/migrating-rails-app-from-heroku-to-google-appengine.md

```diff
@@ -0,0 +1,418 @@
+
+1.  Secure your new database instance by setting a password on the default
+    postgres user:
+
+        gcloud sql users set-password postgres no-host \
+          --instance=mydbinstance --prompt-for-password
+
+      When prompted, enter a password for the database.
+
+1.  Create a database:
+
+        gcloud sql databases create mydb --instance=mydbinstance
+
+Currently, your database is owned by the default user, `postgres`. You can
+change the ownership to an unprivileged user later, but for now you should
+keep the default user to ensure that the data migration goes smoothly.
+
+### Connect to your Cloud SQL instance
+
+By default, Cloud SQL instances are secured: to connect using standard
+Postgres protocols, you are required to whitelist your your IP address.
+This security measure, while necessary, makes it challenging to establish
+_ad hoc_ database connections.
+
+Cloud SQL provides the [Cloud SQL Proxy](https://cloud.google.com/sql/docs/postgres/sql-proxy)
+command-line tool, which opens a database endpoint that you can connect to locally and which
+proxies back to the database instance via a secure API.
+
+To set up Cloud SQL Proxy, perform the following steps:
+
+1.  [Enable the Google Cloud SQL API in the Cloud Console](https://console.cloud.google.com/apis/library/sqladmin.googleapis.com/).
+
+1.  [Install Cloud SQL Proxy](https://cloud.google.com/sql/docs/postgres/sql-proxy#install).
+    Ensure that `cloud_sql_proxy` is in your environment's `PATH` variable.
+
+1.  The proxy creates local sockets connected to your database instances. In your Rails app
+    directory, create a directory to hold those sockets:
+
+        cd /path/to/my/rails/app
+        mkdir -p tmp/cloudsql
+
+1.  To start the proxy, run the following command:
+
+        cloud_sql_proxy -dir=tmp/cloudsql
+
+    Note: This runs the proxy in the foreground, so subsequent commands
+    need to be run in a separate shell. If you prefer, feel free to
+    background the process instead.
+
+5.  The proxy opens a socket in the directory
+    `/path/to/my/rails/app/tmp/cloudsql/[CONNECTION-NAME]/`. You can then
+    connect to the database simply by pointing at that socket. If you like, you
+    can test connecting with `psql:`
+
+        psql -h /path/to/my/rails/app/tmp/cloudsql/[CONNECTION-NAME] \
+          -U postgres
+
+### Copy data from Heroku to Cloud SQL
+
+Now that you have a way to connect directly to your database instance, you can
+copy the data over from your Heroku database.
+
+1.  Export the contents of the Heroku database:
+
+        heroku pg:backups:capture
+        heroku pg:backups:download
+
+    Those commands create a file `latest.dump` in the current working directory.
+
+1.  Restore the data to your Cloud SQL database. Use the socket provided by
+    the proxy to connect to the database instance:
+
+        pg_restore --verbose --clean --no-acl --no-owner -d mydb -U postgres \
```

Comment:
Ohhhh I had no idea you could do that (use `pg_restore` essentially to convert a binary format dump to a text format dump). That might make it unnecessary to use the CloudSQL proxy, which would simplify things tremendously. Lemme work through that. Thanks @frankyn !

Reply:
Found issues with this idea.

After deployment I realized that the UI import had failed because my tables were missing. 

1. My Heroku app uses PostgreSQL 10.1 which is necessary to parse the `latest.dump` `CREATE SEQUENCE` command, but Cloud SQL for PostgreSQL only supports [9.6](https://cloud.google.com/sql/docs/postgres/create-instance). 10.1 IIUC [introduced `AS data_type`](https://www.postgresql.org/docs/current/static/sql-createsequence.html) into the `CREATE SEQUENCE` command and causes the UI import to fail as it's unable to parse. 
**Additional note**: any error will fail the entire import.
**The workaround** is to remove `AS integer`. 

```
CREATE SEQUENCE "widgets_id_seq"
    AS integer << Issue
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;
```

2. Second issue I noticed is the `latest.dump` contains extension `plpgsql` information which user `postgres` and `cloudsqlsuperuser` don't have permission to execute the following SQL commands:
  ```
--
-- Name: plpgsql; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS "plpgsql" WITH SCHEMA "pg_catalog";


--
-- Name: EXTENSION "plpgsql"; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON EXTENSION "plpgsql" IS 'PL/pgSQL procedural language';

Error: SET SET SET SET SET SET SET SET COMMENT CREATE EXTENSION ERROR: must be owner of extension plpgsql This may take a few minutes. While this operation is running, you may continue to view information about the instance.
```
**The workaround** is to remove these commands. Moreover, I did notice that the plpgsql extension exists in the pg_catalog when I reviewed the database extensions using `pgAdmin`.

**Additional note:**
I was able to get `pg_restore` to work with the cloud sql proxy. When I do the restore I see a number of errors related to the issues mentioned above, but the import still seems successful. As the tables and data are available.

Reply:
Hey @frankyn thanks for the additional details.

Ugh... the Postgres 10 vs 9.6 issue is going to be a problem—I didn't realize Heroku [now defaults new database to Postgres 10](https://blog.heroku.com/postgresql-10-general-availability) just as of a couple weeks ago. That might require us to put this on hold for a bit until Cloud SQL gets Postgres 10 support. (I too am running into this issue.) I don't like the thought of having to workaround a bunch of feature changes between major versions, and the `AS` clause in `CREATE SEQUENCE` might not be the only one people will encounter in arbitrary databases.

Setting the postgres version aside for the moment, though... Regarding `pg_restore` using the cloud sql proxy, do you know what your original issue was? Is the command, as listed, incorrect? I'm still trying to weigh the various pros and cons between the cloud sql proxy technique (which requires installing and configuring the proxy), and using the import tool (which requires converting the dump, possibly editing it to work around the `CREATE EXTENSION` issue, uploading to GCS, and setting a bunch of permissions).

Reply:
IIRC, the issue was I missed the `latest.dump` part of the command and it would hang. I misunderstood this state as it was attempting to connect. 

I did however hit a maximum byte length of 103 when including my home directory as the `mkdir -p tmp/cloudsql` didn't state where I should place this directory. 

In response to your pros and cons statement. There still exists an issue with 10.1 != 9.6. When I use the cloud sql proxy, I was able to import the 10.1 SQL file with errors, and it was for a simple getting started project. Not sure how this may impact more complex projects.